marshal/admin.go

```diff
@@ -0,0 +1,242 @@
+/*
+ * portal - marshal
+ *
+ * a library that implements an algorithm for doing consumer coordination within Kafka, rather
+ * than using Zookeeper or another external system.
+ *
+ */
+
+package marshal
+
+import (
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/dropbox/kafka/proto"
+)
+
+const (
+	consumerReleaseClaimWaitSleep = 5 * time.Second
+	consumerReleaseClaimWaitTime  = 15 * time.Minute
+)
+
+// Admin is used to pause a consumer group and set what position it reads from
+// for certain partitions.
+type Admin interface {
+	// SetConsumerGroupPosition resets consumers to read starting from
+	// offsets on each topic, partition pair in positions.
+	SetConsumerGroupPosition(groupID string, offsets map[string]map[int]int64) error
+}
+
+type consumerGroupAdmin struct {
+	clientID     string
+	groupID      string
+	marshaler    *Marshaler
+	pauseTimeout time.Duration
+	steps        *int32
+}
+
+func (m *Marshaler) NewAdmin(groupID string, pauseTimeout time.Duration) (Admin, error) {
+	if len(m.consumers) != 0 {
```

Comment:
You can't access member variables without locks. Can you make sure that you're careful about the locking semantics in these methods? Most of the lock requirements are documented on the structs, I believe, but in general you have to be careful about what you access.